service/src/main/java/com/epam/esm/services/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,197 @@
+package com.epam.esm.services.service.impl;
+
+import com.epam.esm.model.entity.Entity;
+import com.epam.esm.model.entity.Tag;
+import com.epam.esm.persistence.util.EntityFinder;
+import com.epam.esm.persistence.util.CertificateFinder;
+import com.epam.esm.persistence.dao.impl.CertificateDAOImpl;
+import com.epam.esm.model.entity.Certificate;
+import com.epam.esm.persistence.exceptions.DAOSQLException;
+import com.epam.esm.services.exceptions.BadRequestException;
+import com.epam.esm.services.exceptions.ServiceException;
+import com.epam.esm.services.exceptions.ValidationException;
+import com.epam.esm.services.constants.CertificateSearchParameters;
+import com.epam.esm.services.constants.CertificateSortingParameters;
+import com.epam.esm.services.service.CertificateService;
+import com.epam.esm.services.validator.EntityValidator;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private CertificateDAOImpl dao;
```

Comment:
you should use interfaces instead of implementation

Reply:
Done.